uni/lib/model/providers/startup/session_provider.dart

```diff
@@ -111,4 +123,83 @@ class SessionProvider extends StateProviderNotifier<Session> {
 
     await acceptTermsAndConditions();
   }
+
+  late Flow? _flow;
+  late material.BuildContext? context;
+  bool _persistentSession = false;
+
+  Future<void> federatedAuthentication(
+    material.BuildContext ctx, {
+    required bool persistentSession,
+  }) async {
+    context = ctx;
+    _persistentSession = persistentSession;
+
+    final realm = dotenv.env['REALM'] ?? '';
+    final issuer = await Issuer.discover(Uri.parse(realm));
+    final client = Client(
+      issuer,
+      dotenv.env['CLIENT_ID']!,
+    );
+
+    _flow = Flow.authorizationCodeWithPKCE(
+      client,
+      scopes: [
+        'openid',
+        'profile',
+        'email',
+        'offline_access',
+        'audience',
+        'uporto_data',
+      ],
+    );
+    _flow?.redirectUri = Uri.parse('pt.up.fe.ni.uni://auth');
+
+    await _invoke(_flow!.authenticationUri);
+  }
+
+  Future<void> finishFederatedAuthentication(Uri uri) async {
+    final credential = await _flow!.callback(uri.queryParameters);
+    final userInfo = (await credential.getUserInfo()).toJson();
+    final token = (await credential.getTokenResponse()).accessToken;
```

Comment:
This logic is a bit too complex to be on the provider. Maybe the late variables can be avoided with a slight refactor.

Reply:
Know that you can say that, I think I can move up the flow instance creation. But the provider stills looks to me the place to stay that since the job is to update the session. Any suggestions on that?